### DISCUSSION TITLE
crypto/promises

### DISCUSSION THREAD

**Discussion by panva** [nodejs/help#4825] [Q&A]
Now that crypto internals are refactored and Web Cryptography API landed, i think it's time to put `crypto/promises` on the roadmap so that we can push APIs like `crypto.sign`, `crypto.verify`, `crypto.diffieHellman`, `crypto.publicEncrypt`, `crypto.privateDecrypt`, `Cipher.update/final` and `Decipher.update/final` to the libuv threadpool.

@jasnell WDYT?

**Comment by Trott** [comment]
@nodejs/crypto

**Comment by jasnell** [comment]
Certainly Possible. If you'd like to start working on that I'll definitely support through reviews

**Reply by panva** [reply]
If this was a pure JS job i'd be all over it long time ago. However, looking at the bindings difference between e.g. signOneShot and webcrypto sign I am getting confused. I will give it a try though.

**Reply by panva** [reply]
@jasnell it is, sadly, beyond my abilities to make this work. I've set out to start with 'crypto/promises'.sign(algorithm, key, data). I have a SignJob executing and returning a buffer, it uses `preparePrivateKey` to process the key argument, but its resulting format, type, passphrase has no place in the SignJob constructor, same goes for the DSA Signature Encoding option.

**Reply by jasnell** [reply]
Ok, it'll likely be a week or two before I can get back to this but I'll start taking a look
